Add support for NodePort when exposing deployment

masap opened the pull request:
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Previously, ```expose``` command does not allow to specify ```--node-port``` even when the type of service is NodePort.

```
$ kubectl create deployment test-deploy --image=nginx
$ kubectl expose deployment.apps/test-deploy --port=80 --type=NodePort --node-port=30082
Error: unknown flag: --node-port
See 'kubectl expose --help' for usage.
```

**Which issue(s) this PR fixes**:
NONE

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Add support for NodePort when exposing deployment
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

masap commented:
/ok-to-test

masap commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

dougsland commented:
/cc

eddiezane commented on staging/src/k8s.io/kubectl/pkg/cmd/expose/expose.go:
```diff
@@ -289,6 +291,11 @@ func (o *ExposeServiceOptions) RunExpose(cmd *cobra.Command, args []string) erro
 
+		// Only type=NodePort supports --node-port
+		if o.NodePort != 0 && params["type"] != "NodePort" {
+			return cmdutil.UsageErrorf(cmd, "couldn't specify --node-port when type is not NodePort")
```
nit: s/couldn't/cannot

masap replied:
Thanks ! Fixed.

eddiezane commented on staging/src/k8s.io/kubectl/pkg/generate/versioned/service.go:
```diff
@@ -210,6 +211,11 @@ func generateService(genericParams map[string]interface{}) (runtime.Object, erro
 			service.Spec.Ports[i].TargetPort = intstr.FromInt(int(port))
 		}
 	}
+	if nodePort, err := strconv.Atoi(params["node-port"]); err == nil {
```
Let's return this error.

masap replied:
Fixed as well.

masap commented:
/test pull-kubernetes-e2e-kind-ipv6

dougsland commented on staging/src/k8s.io/kubectl/pkg/generate/versioned/service.go:
```diff
@@ -210,6 +211,15 @@ func generateService(genericParams map[string]interface{}) (runtime.Object, erro
 			service.Spec.Ports[i].TargetPort = intstr.FromInt(int(port))
 		}
 	}
+	if params["node-port"] != "" {
```
I am wondering if shouldn't be:
```
if len(params["node-port"]) != 0 {
```

masap replied:
Thanks! Yes, existing lines are written with such a manner. Fixed.

dougsland commented on staging/src/k8s.io/kubectl/pkg/cmd/expose/expose.go:
```diff
@@ -154,6 +155,7 @@ func NewCmdExposeService(f cmdutil.Factory, streams genericclioptions.IOStreams)
 	cmd.Flags().String("container-port", "", i18n.T("Synonym for --target-port"))
 	cmd.Flags().MarkDeprecated("container-port", "--container-port will be removed in the future, please use --target-port instead")
 	cmd.Flags().String("target-port", "", i18n.T("Name or number for the port on the container that the service should direct traffic to. Optional."))
+	cmd.Flags().IntVar(&o.NodePort, "node-port", o.NodePort, "Port used to expose the service on each node in a cluster.")
```
I would suggest add i18n.T:
```
i18n.T("Name or number..
```

masap replied:
Thanks, fixed. I have copied this line from ```create``` command, so I fixed it as well.

https://github.com/kubernetes/kubernetes/blob/3a22c023b9f1a7344ad96343cd68bc7409aee4ba/staging/src/k8s.io/kubectl/pkg/cmd/create/create_service.go#L297

dougsland replied:
@masap different PR ? What's the #?

masap replied:
It is in this PR.

masap replied:
https://github.com/kubernetes/kubernetes/pull/98244/files#diff-332302c0f142f29cb2425c6e44701022fcaf75515e8ef04181dfa0fc1e5fa8a5

masap commented:
/test pull-kubernetes-typecheck

masap commented:
/retest

dougsland commented:
/lgtm

masap commented:
/assign @mengqiy

eddiezane commented:
I just noticed that `port` and `target-port` are both treated as string parameters. Is there a reason to treat this differently?

masap commented:
```target-port``` is treated as string parameters, because it could receive a string like ```--target-port=http```.
On the other hand, ```port``` and ```node-port``` receive only a number.
This is an excerpt of ```kubectl explain service --recursive```.
```
      ports	<[]Object>
         nodePort	<integer>
         port		<integer>
         targetPort	<string>
```

aojea commented on staging/src/k8s.io/kubectl/pkg/generate/versioned/service.go:
```diff
@@ -210,6 +211,15 @@ func generateService(genericParams map[string]interface{}) (runtime.Object, erro
+			return nil, fmt.Errorf("invalid node-port: %s", params["node-port"])
+		}
+		for i := range service.Spec.Ports {
+			service.Spec.Ports[i].NodePort = int32(nodePort)
```
you can not use the same nodeport for different ports

masap replied:
@aojea Thanks! Indeed it fails when 2 port numbers are specified.
```
$ kubectl create deployment test-deploy --image=nginx
$ kubectl expose deployment.apps/test-deploy --type=NodePort --port=80,81 --node-port=30080
The Service "test-deploy" is invalid: spec.ports[1].nodePort: Duplicate value: 30080
```
So, I fixed ```--node-port``` parameter to be able to receive multiple port numbers like ```--port``` parameter.
```
$ kubectl expose deployment.apps/test-deploy --type=NodePort --port=80,81 --node-port=30080,30081
```

And now I found a similar issue on ```--target-port```, I will fix it after this PR.

@eddiezane Sorry, your insight is great. NodePort should be string to receive multiple port numbers.

masap commented:
/test pull-kubernetes-node-e2e

aojea commented on staging/src/k8s.io/kubectl/pkg/cmd/expose/expose_test.go:
```diff
@@ -630,6 +630,90 @@ func TestRunExposeService(t *testing.T) {
+							Protocol:   corev1.ProtocolUDP,
+							Port:       82,
+							TargetPort: intstr.FromInt(82),
+							NodePort:   31082,
```
the port will be randomly allocated by the apiserver if empty

aojea replied:
in this case I don't think that the cli should be smart and assume the user wants the net, it may happen the nodeports are not consecutively.
I think that it should be left empty or just fail, maybe fail since the user is specifying node-ports explicitly, so we can assume he forgot to set it, otherwise is not normal to set only ones and left others

masap replied:
On ```kubectl create service```, if target port is absent it is assumed to be default (same as port).
```
$ kubectl create service nodeport test-nodeport --tcp=80:8080 --tcp=81 --dry-run=client -o yaml
apiVersion: v1
kind: Service
metadata:
  creationTimestamp: null
  labels:
    app: test-nodeport
  name: test-nodeport
spec:
  ports:
  - name: 80-8080
    port: 80
    protocol: TCP
    targetPort: 8080
  - name: "81"
    port: 81
    protocol: TCP
    targetPort: 81
  selector:
    app: test-nodeport
  type: NodePort
status:
  loadBalancer: {}
```
So I think it is natural that node port assume default when it is absent.
How do you think ?

aojea replied:
yeah, but I refer to this case
```
, "port": "80,81,82",  "node-port": "30080,30081"
```
I don't see that is correct default to 30082, what happens if node-port is: "30080,31223", do you use 31224?
Also, nodeports can be set by the user or otherwise they are randomly allocated by the apiserver, so the default should be to left it empty.
But my point is that in case that the user specify multi-ports, we should validate that the number of ports and node-port match, because the user is explicitly saying he wants "those" multi-ports, so it feels weird to say I want this specific nodeports but just giveme some other random :)

masap replied:
> I don't see that is correct default to 30082, what happens if node-port is: "30080,31223", do you use 31224?

Do not worry about it, I understand the third node-port value will be assinged randomly.
I just use 31082 because I need some number.
I will remove this case, but I want to know how to write test case for such a random value.

> But my point is that in case that the user specify multi-ports, we should
> validate that the number of ports and node-port match, because the user is
> explicitly saying he wants "those" multi-ports, so it feels weird to say I
> want this specific nodeports but just giveme some other random :)

Ok, I will fix this.

aojea commented on staging/src/k8s.io/kubectl/pkg/generate/versioned/service.go:
```diff
@@ -210,6 +211,19 @@ func generateService(genericParams map[string]interface{}) (runtime.Object, erro
 	}
+	if len(params["node-port"]) > 0 {
+		nodePortStringSlice := strings.Split(params["node-port"], ",")
+		if len(service.Spec.Ports) < len(nodePortStringSlice) {
```
I think that this should work only if the node-ports match the number of ports https://github.com/kubernetes/kubernetes/pull/98244/files#r582755570
```suggestion
		if len(service.Spec.Ports) != len(nodePortStringSlice) {
```

masap replied:
Thanks, fixed.

masap commented:
/test pull-kubernetes-node-e2e

masap committed:
Add support for NodePort when exposing deployment

Previously, expose command does not allow to specify --node-port even
when the type of service is NodePort.

$ kubectl create deployment test-deploy --image=nginx
$ kubectl expose deployment.apps/test-deploy --port=80 --type=NodePort --node-port=30082
Error: unknown flag: --node-port
See 'kubectl expose --help' for usage.

aojea commented:
/lgtm
Thanks

soltysh commented on staging/src/k8s.io/kubectl/pkg/cmd/expose/expose.go:
```diff
@@ -154,6 +154,7 @@ func NewCmdExposeService(f cmdutil.Factory, streams genericclioptions.IOStreams)
 	cmd.Flags().String("container-port", "", i18n.T("Synonym for --target-port"))
 	cmd.Flags().MarkDeprecated("container-port", "--container-port will be removed in the future, please use --target-port instead")
 	cmd.Flags().String("target-port", "", i18n.T("Name or number for the port on the container that the service should direct traffic to. Optional."))
+	cmd.Flags().String("node-port", "", i18n.T("Port used to expose the service on each node in a cluster."))
```
From what I see `kubectl expose` is following the path that we stopped `kubectl run` from. Iow. it's growing its functionality beyond basic capabilities. There is `kubectl create nodeport` for advanced use cases and I'd prefer we don't support all possibilities here.

soltysh replied:
Additionally, we're on a path to get rid of generators of any flavor so this PR would be stomping that effort along the way too.

masap replied:
> From what I see kubectl expose is following the path that we stopped kubectl run from.

I have no problem with closing this PR.
Are there any documents that I should read before I start development that indicate such a development policy?

soltysh reviewed (changes requested):
/hold
/lgtm cancel

aojea commented:
> /hold
> /lgtm cancel

my bad, sorry @soltysh I wrongly assumed this was already triaged and missing the networking bits

soltysh commented:
Closing per author agreement.
/close
